paddle/fluid/operators/center_loss_op.cc

```diff
@@ -0,0 +1,152 @@
+
+#include "paddle/fluid/operators/center_loss_op.h"
+#include <memory>
+#include <string>
+
+namespace paddle {
+namespace operators {
+class CenterLossOp : public framework::OperatorWithKernel {
+ public:
+  CenterLossOp(const std::string &type,
+               const framework::VariableNameMap &inputs,
+               const framework::VariableNameMap &outputs,
+               const framework::AttributeMap &attrs)
+      : OperatorWithKernel(type, inputs, outputs, attrs) {}
+
+  void InferShape(framework::InferShapeContext *ctx) const override {
+    PADDLE_ENFORCE(ctx->HasInput("X"),
+                   "Input(X) of CenterLoss should not be null.");
+    auto x_dims = ctx->GetInputDim("X");
+
+    PADDLE_ENFORCE(ctx->HasInput("CenterUpdateRate"),
+                   "Input(CenterUpdateRate) of CenterLoss should not be null.");
+
+    PADDLE_ENFORCE(ctx->HasInput("Label"),
+                   "Input(Label) of CenterLoss should not be null.");
+
+    PADDLE_ENFORCE(ctx->HasInput("Centers"),
+                   "Input(Centers) of CenterLoss should not be null.");
+
+    PADDLE_ENFORCE(ctx->HasOutput("CentersDiff"),
+                   "Output(CentersDiff) of CenterLoss should not be null.");
+
+    PADDLE_ENFORCE(ctx->HasOutput("Loss"),
+                   "Output(Loss) of CenterLoss should not be null.");
+
+    PADDLE_ENFORCE(
+        ctx->HasOutput("CentersOut"),
+        "Output(CentersOut) of CenterLoss shared data with Centers.");
+
+    ctx->SetOutputDim("CentersDiff", {x_dims[0], product(x_dims) / x_dims[0]});
+    ctx->SetOutputDim("CentersOut", ctx->GetInputDim("Centers"));
+    ctx->SetOutputDim("Loss", {x_dims[0], 1});
+    ctx->ShareLoD("X", /*->*/ "Loss");
+  }
+
+ protected:
+  framework::OpKernelType GetExpectedKernelType(
+      const framework::ExecutionContext &ctx) const override {
+    return framework::OpKernelType(ctx.Input<Tensor>("X")->type(),
+                                   ctx.device_context());
+  }
+};
+
+class CenterLossOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  void Make() override {
+    AddInput("X", "(Tensor) Input tensor of center_loss operator.");
+    AddInput("Label", "(Tensor) Input tensor of center_loss operator.");
+    AddInput("Centers", "(Tensor) Input tensor of center_loss operator.");
+    AddInput("CenterUpdateRate",
+             "(Tensor) Input tensor of center_loss operator.");
+
+    AddOutput("CentersOut", "(Tensor) Input tensor of center_loss operator.");
+    AddOutput("CentersDiff", "(Tensor) output tensor of center_loss operator.");
+    AddOutput("Loss", "(Tensor) Output tensor of center_loss operator.");
+
+    AddAttr<int>("cluster_num",
+                 "The output cluster num of the center_loss operator.");
+    AddAttr<bool>("need_update", "whether need to update center info.");
+    AddComment(R"DOC(
+**CenterLoss operator**
+implemention of the algorithm in the papper A Discriminative 
+Feature Learning Approach for Deep Face Recognition
```

Comment:
the implementation is just a loss, not the full algorithm introduced in that paper. It is better to describe what the loss is and what the feature is (or rather what we use this loss for).

Reply:
done